feat: Deploy applications with custom resources

gatici opened the pull request:
## Description

This PR provides to use custom, local charm resources with applications. The application resource could be a charm revision or another resource which needs to be uploaded or pulled from a repository. The following "juju_application" resource definition becomes valid within this PR.

```
resource "juju_application" "ausf" {
  name = "ausf"
  model = var.model_name

  charm {
    name = "sdcore-ausf-k8s"
    channel = var.channel
  }
  resources = {
      ausf-image = "gatici/sdcore-ausf:1.4.0"
      first-image =  0
      second-image =  "3"
      third-resource =  "/home/gatici/myimage.json"
      fourth-resource =  "/home/gatici/myimage.yaml"
  }
  units = 1
  trust = true
}
```


Fixes: https://github.com/juju/terraform-provider-juju/issues/460

## Type of change

- Change existing resource
- Change in tests (one or several tests have been changed)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Requires a documentation update

## Environment
Tested with following versions.

- Juju controller version: 3.4


- Terraform version:  0.12.0


## QA steps

Manual QA steps should be done to test this PR.

```tf
provider juju {}
...
```

## Additional notes

hmlanigan reviewed (changes requested):
Thank you for your contribution. It's a good start.

 To proceed, please read and follow the process defined: https://github.com/juju/terraform-provider-juju/wiki/Developing#planning--design

Please also include description of what happens during the following scenarios.
- deploy
- refresh
- the resource element is removed from the plan
- what happens if the file isn't an image location?
- what happens if the value is changed from an int to a file?
- what happens if the value is changed from a file to an int?

gatici commented:
> Thank you for your contribution. It's a good start.
> 
> To proceed, please read and follow the process defined: https://github.com/juju/terraform-provider-juju/wiki/Developing#planning--design
> 
> Please also include description of what happens during the following scenarios.
> 
>     * deploy
> 
>     * refresh
> 
>     * the resource element is removed from the plan
> 
>     * what happens if the file isn't an image location?
> 
>     * what happens if the value is changed from an int to a file?
> 
>     * what happens if the value is changed from a file to an int?

Hello @hmlanigan,
The behaviour is summarized below. Please tell me if there is something undesired.
```
  * If the plan does not specify a resource and resources are added to the plan (as a revision number or a custom resource), specified resources are attached to the application (equivalent to juju attach-resource).
  * If the plan does specify resource revisions and if the charm revision or channel is updated, existing resources are kept.  (Resources are not detached)
  * If the plan does specify resource revisions and resources are removed from the plan:
          - If charm revision/channel is updated, the resources associated with the updated charm revision or channel  is attached.
          - If the charm revision/channel are not updated then the resources associated with the existing charm revision/channel are attached.
  * Charm could be deployed without resource, then resource could be  added later.
  * Resources could be provided in the following formats:
          - A custom repository URL
          - A files ends with .txt, .json  or .yaml
          - A resource revision from CharmHub
    * Charm could be deployed with resources.
    * If the provided resource revision does not exist during initial deployment or update, Client does not start deployment with an error that resource was not found in the store.
    * If the provided custom resource does not exist during initial deployment or update, Client  start deployment and charm could not be deployed properly and charm will be in error state.
    *  If the Provided resource type is not correct then Client fails with incorrect resource error for below scenarios:
          - An image is expected but file does not include image URL
          - A plain text file is expected but a .json file is provided.
    *  If the provided resource does not exist then Client fails with path is not valid error.
    *  If provided resource value is changed from an int to a file  then the image resource from the file is attached.
    *  If provided resource value is changed from a file to an int then image revision is attached.
```

hmlanigan commented:
> * If the plan does specify resource revisions and if the charm revision or channel is updated, existing resources are kept.  (Resources are not detached)
 
This is counter to current juju behavior. Today juju will upgrade the resources with the charm if a new resource is available, provided that the resource is not an attached local file (but not an ici-image). 

This behavior would have to be explicitly implemented if we want it. 

Resources can never be detached by juju.

hmlanigan commented:
>   * If the plan does specify resource revisions and resources are removed from the plan:
>           - If charm revision/channel is updated, the resources associated with the updated charm revision or channel  is attached.
>           - If the charm revision/channel are not updated then the resources associated with the existing charm revision/channel are attached.

A charm revision may be in more than one channel. The last I knew resources are associated with a channel only today, not a charm revision.

gatici commented:
The review comments are addressed to update the resources from latest revision associated with the channel upon charm channel updates.  As summarized below, only image resources could be provided as custom resource inputs. Image information could be written directly in a string format or provided in a json or yaml file. The PR is compatible with both works with deployFromRepository and legacyDeploy methods and older Juju versions are supported.

```
Charms could be deployed with specifying the resources (from CharmHub or an OCI image repository). If the resources are not specified, the resources which are associated with the Charm in the CharmHub are used. Resource inputs are provided in a string format.
-         Resource revision number from CharmHub (string)
-         OCI image information as a URL (string)
-         A path of json or yaml file which includes OCI image repository information (string)

Changing resource input from a revision to a custom OCI resource is processed and updated smoothly according to the provided input.

Resources could be added to the Terraform plan after deployment.
-         If the resources are added to the plan (as a revision number or a custom OCI image resource), specified resources are attached to the application (equivalent to juju attach-resource).

Charm which includes resources could be updated. If the plan does specify resource revisions from CharmHub:
-         if the charm channel is updated, resources get updated to the latest revision associated with the updated channel. If the plan does specify custom OCI image resources:
-         if the charm channel is updated, existing resources are kept. (Resources are not detached)

Resources could be removed from the Terraform plan. If the plan does specify resource revisions from CharmHub or custom OCI images, then resources are removed from the plan:
-         If the charm channel is updated, resources get updated to the latest revision associated with the updated charm channel.
-         If the charm channel is not updated then the resources get updated to the latest revision associated with the existing charm channel.
```

hmlanigan reviewed (commented):
I'm part way through my review. However I've run out of time to complete it, this is what I have now. I had to spend some time refreshing my knowledge of how juju handles resources, it's a complex bit of code.

I see no need to add the json or yaml file option for oci-images as they can be directly listed in the plan as a URL. It also muddies the water around uploading file resources which we do not intend to support.

gatici commented on examples/resources/juju_application/resource.tf:
```diff
@@ -33,4 +33,22 @@ resource "juju_application" "placement_example" {
   }
+}
+
+resource "juju_application" "custom_resources_example" {
```
OK, Done.

gatici commented on internal/provider/validator_base.go:
```diff
@@ -7,7 +7,6 @@ import (
 	"context"
 
 	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
-
```
Done

gatici commented on internal/provider/resource_application.go:
```diff
@@ -28,7 +29,6 @@ import (
 	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/juju/errors"
 	"github.com/juju/juju/core/constraints"
-
```
Done

gatici commented on internal/provider/resource_application_test.go:
```diff
@@ -12,13 +12,11 @@ import (
 	apispaces "github.com/juju/juju/api/client/spaces"
 	"github.com/juju/juju/rpc/params"
 	"github.com/juju/names/v5"
-
```
Added again. Done

gatici commented on internal/provider/resource_application.go:
```diff
@@ -882,29 +901,65 @@ func (r *applicationResource) Update(ctx context.Context, req resource.UpdateReq
+			for k, v := range updateApplicationInput.Resources {
+				if isInt(v) {
+					// Set resource revision to zero gets the latest resource revision from CharmHub
+					updateApplicationInput.Resources[k] = "0"
```
When we list the charm resources of any charm, I can not see any revision number  equals to 0.  
Where and how the revision zero is used ?
For example:
 $ charmcraft  resources  sdcore-udm-k8s                                                                                                                                                                                    
9            udm-image   oci-image  True                                                                                                                                                                                       
8            udm-image   oci-image  True                                                                                                                                                                                       
7            udm-image   oci-image  True                                                                                                                                                                                       
6            udm-image   oci-image  True                                                                                                                                                                                       
5            udm-image   oci-image  True                                                                                                                                                                                       
4            udm-image   oci-image  True                                                                                                                                                                                       
3            udm-image   oci-image  True                                                                                                                                                                                       
2            udm-image   oci-image  True                                                                                                                                                                                       
1            udm-image   oci-image  True

gatici replied:
OK, -1 is used  to indicate no revision for ints.

gatici commented on internal/provider/resource_application.go:
```diff
@@ -43,17 +43,25 @@ const (
-		  latest revision.
-	    * If the charm revision or channel are not updated, then no changes will take 
-		  place (juju does not have an "un-attach" command for resources).
+Charm resources. Must evaluate to a string. A resource could be a resource revision number from CharmHub or a custom OCI image resource.
+There are a few scenarios that need to be considered:
+* Charms could be deployed with specifying the resources (from CharmHub or an OCI image repository). If the resources are not specified, the resources which are associated with the Charm in the CharmHub are used.
+  Resource inputs are provided in a string format.
+  - Resource revision number from CharmHub (string)
+  - OCI image information as a URL (string)
+  - A path of json or yaml file which includes OCI image repository information (string)
+* Changing resource input from a revision to a custom OCI resource is processed and updated smoothly according to the provided input.
+* Resources could be added to the Terraform plan after deployment.
+  - If the resources are added to the plan (as a revision number or a custom OCI image resource), specified resources are attached to the application (equivalent to juju attach-resource).
+* Charm which includes resources could be updated.
+  If the plan does specify resource revisions from CharmHub:
+  - if the charm channel is updated, resources get updated to the latest revision associated with the updated channel.
+  If the plan does specify custom OCI image resources:
+  - if the charm channel is updated, existing resources are kept. (Resources are not detached)
+* Resources could be removed from the Terraform plan.
+  If the plan does specify resource revisions from CharmHub or custom OCI images, then resources are removed from the plan:
+  - If the charm channel is updated, resources get updated to the latest revision associated with the updated charm channel.
+  - If the charm channel is not updated then the resources get updated to the latest revision associated with the existing charm channel.
```
`
Resources specified by URL to an OCI image repository will never be refreshed (upgraded) by juju during a charm refresh unless explicitly changed in the plan. This includes resources that were once specified by URL and changed to a revision number.
`
I have some doubts with these sentences.  
We say Juju refresh will not refresh attached OCI images. I totally agree.  
Second sentence tells that resources specified by URL and changed to a revision number coincides with the condition of 
explicitly changed in the plan.  Hence, resource is expected to be refreshed.

Please check the suitability of following summary and confirm.
`
OCI image is already  attached --->  channel is refreshed  -->  No resource change

OCI image already attached  -->  resource is changed to revision  and  channel refreshed  -- > resource is refreshed according to channel

OCI image already attached  -->   resource is changed to revision   -- > resource is refreshed according to revision

OCI image already attached -->  resource is changed to None -->  resource is refreshed according to current channel

Resource is revision numver -->   resource is changed to None  --> resource is refreshed according to current channel
`

gatici replied:
https://docs.google.com/document/d/1i236ntmw-qXyqifmEtk_rcszWXcXSUhXkeJ6ByKXIeE/edit is written to come to an agreement  about the behavior.

gatici commented on internal/provider/resource_application.go:
```diff
@@ -265,7 +284,7 @@ func (r *applicationResource) Schema(_ context.Context, _ resource.SchemaRequest
 			ResourceKey: schema.MapAttribute{
 				Optional:            true,
-				ElementType:         types.Int64Type,
+				ElementType:         types.StringType,
```
This is not a breaking change. Existing plans may include Int64 and using using String type will store the values as string. The upgrades could be done smoothly. Tested and verified manually.

gatici commented on internal/provider/resource_application.go:
```diff
@@ -265,7 +284,7 @@ func (r *applicationResource) Schema(_ context.Context, _ resource.SchemaRequest
 			},
 			ResourceKey: schema.MapAttribute{
```
Added.

gatici commented on internal/juju/applications.go:
```diff
@@ -1350,45 +1468,176 @@ func (c applicationsClient) updateResources(appName string, resources map[string
 }
 
-func addPendingResources(appName string, resourcesToBeAdded map[string]charmresources.Meta, resourceRevisions map[string]int,
+func addPendingResources(appName string, resourcesToBeAdded map[string]charmresources.Meta, resourcesRevisions map[string]string,
```
Function arguments are renamed.

gatici replied:
Function arguments are renamed.

hmlanigan commented on internal/provider/resource_application.go:
```diff
@@ -961,29 +972,41 @@ func (r *applicationResource) Update(ctx context.Context, req resource.UpdateReq
 	// NOT to update resources, because we want revisions fixed to those
```
nit: please update this line of the comments, we're no longer focused solely on revisions.

gatici replied:
Done

hmlanigan commented on internal/provider/validator_resourcekey.go:
```diff
@@ -0,0 +1,71 @@
+			imageUrlPattern := `(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]):[\w][\w.-]{0,127}`
+			urlRegex := regexp.MustCompile(imageUrlPattern)
+			if urlRegex.MatchString(value) {
+				return
```
issue: return or continue? What happens if there is more than one oci-image resource provided?  Is that possible? What happens if the charm has an oci-image and file resources?

In general this method requires more comments about what is going on around the differences in file and oci-image resources in the validation.

gatici replied:
More than one OCI image could be provided. Continue is used.

hmlanigan commented on internal/provider/validator_resourcekey.go:
```diff
@@ -0,0 +1,71 @@
+			urlRegex := regexp.MustCompile(imageUrlPattern)
+			if urlRegex.MatchString(value) {
+				return
+			} else {
```
polish: else isn't required here.

gatici replied:
Done

hmlanigan commented on internal/provider/validator_resourcekey.go:
```diff
@@ -0,0 +1,71 @@
+					req.Path,
+					"Invalid Resource revision",
+					fmt.Sprintf("value of %q should be a valid revision number or image URL: %s", name, err),
+				)
```
todo: add a `continue` after AddAttributeError, no need to check the value.

gatici replied:
Done

hmlanigan commented on internal/provider/validator_resourcekey.go:
```diff
@@ -0,0 +1,71 @@
+					"Invalid Resource revision",
+					fmt.Sprintf("value of %q should be a valid revision number or image URL: %s", name, "Negative revision number is invalid."),
+				)
+			}
+		} else {
```
todo: if you use a continue, the else isn't needed with the extra indentation.

gatici replied:
Done

hmlanigan commented on internal/provider/validator_resourcekey_test.go:
```diff
@@ -0,0 +1,12 @@
+	"testing"
+)
+
+func TestResourceKeyValidatorValid(t *testing.T) {
+
+}
```
issue: implement me

gatici replied:
Tests are added.

hmlanigan commented on internal/provider/validator_resourcekey.go:
```diff
@@ -0,0 +1,71 @@
+				resp.Diagnostics.AddAttributeError(
+					req.Path,
+					"Invalid image URL",
+					fmt.Sprintf("value of %q should be a valid revision number or image URL: %s", name, "The value format is invalid as a revision number or for an image URL."),
```
polish: this attribute error is repetitive, can it be simplified?

gatici replied:
Done

hmlanigan commented on internal/provider/resource_application_test.go:
```diff
@@ -344,6 +344,219 @@ func TestAcc_ResourceRevisionUpdatesMicrok8s(t *testing.T) {
 	})
 }
 
+func TestAcc_CustomResourcesAddedToPlanMicrok8s(t *testing.T) {
```
praise: nice tests here

hmlanigan reviewed (commented):
Here are my current comments, though I haven't completed the review yet. 

Mostly polish and nitpicks with 2 main questions/concerns. One is about not using the juju api resource client directly and the other is related to the validator logic. More detail with the lines. 

Some of the changes are to make the code more consistent with already existing code style.

P.S I'm trying out using  https://conventionalcomments.org

hmlanigan commented on internal/provider/resource_application.go:
```diff
@@ -46,17 +48,16 @@ const (
-		  latest revision.
-	    * If the charm revision or channel are not updated, then no changes will take 
-		  place (juju does not have an "un-attach" command for resources).
+Charm resources. Must evaluate to a string. A resource could be a resource revision number from CharmHub or a custom OCI image resource.
+There are a few scenarios that need to be considered:
+
+* Specify a resource other than the default for a charm. Note that not all charms have resources. A resource can be specified by a revision number or by URL to a OCI image repository. Resources of type 'file' can only be specified by revision number. Resources of type 'oci-image' can be specified by revision number or URL.
```
suggestion: When combined with the autogenerated application resource documentation, this reads a bit awkwardly. 

```
Specify a resource other than the default for a charm. Note that not all charms have resources.

Notes:

 *  A resource can be specified by a revision number or by URL to a OCI image repository. Resources of type 'file' can only be specified by revision number. Resources of type 'oci-image' can be specified by revision number or URL to an OCI image repository.
 ```

gatici replied:
Explanation and formatting is updated.

hmlanigan reviewed (commented):
QA went well with both k8s and machine charms and their resources.

gatici commented on internal/provider/validator_resourcekey.go:
```diff
@@ -0,0 +1,71 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
```
Done

gatici commented on internal/provider/validator_resourcekey_test.go:
```diff
@@ -0,0 +1,12 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
```
Done

gatici commented on internal/juju/resources.go:
```diff
@@ -0,0 +1,219 @@
+package juju
```
Added.

gatici commented on internal/provider/resource_application.go:
```diff
@@ -29,6 +30,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/juju/errors"
 	"github.com/juju/juju/core/constraints"
+
```
Done

gatici commented on internal/juju/resources_test.go:
```diff
@@ -0,0 +1,23 @@
+package juju
```
Will be added.

gatici commented on internal/juju/applications.go:
```diff
@@ -391,14 +388,26 @@ func (c applicationsClient) deployFromRepository(applicationAPIClient *apiapplic
+
+	fileSystem := osFilesystem{}
+	// Upload the provided local resources to Juju
+	uploadErr := uploadExistingPendingResources(deployInfo.Name, localPendingResources, fileSystem, resourceHttpClient)
+
+	if uploadErr != nil {
+		return uploadErr
+	}
+	return nil
```
Code is refactored. This method does not exist anymore.

gatici commented on internal/juju/resources.go:
```diff
@@ -0,0 +1,219 @@
+}
+
+// ResourceHttpClient returns a new Client for the given raw API caller.
+func ResourceHttpClient(apiCaller base.APICallCloser) *HttpRequestClient {
```
This is removed.

gatici commented on internal/juju/applications.go:
```diff
@@ -347,8 +348,9 @@ func (c applicationsClient) CreateApplication(ctx context.Context, input *Create
 	}
 
 	applicationAPIClient := apiapplication.NewClient(conn)
+	resourceHttpClient := ResourceHttpClient(conn)
```
ResourceHttpClient is removed.

gatici commented:
> Here are my current comments, though I haven't completed the review yet.
> 
> Mostly polish and nitpicks with 2 main questions/concerns. One is about not using the juju api resource client directly and the other is related to the validator logic. More detail with the lines.
> 
> Some of the changes are to make the code more consistent with already existing code style.
> 
> P.S I'm trying out using https://conventionalcomments.org

Hello Heather,

I have addressed some of the comments but some of them are still  open. I will  continue on Monday. 
As you suggested in your review, a new added client ResourceHttpClient is removed.  I figured out how to use Juju Resource API client for initial deployment stage.  This required some additional info from charm and some helper methods are added under resources.go.

Regarding with the validator, I appreciate for your comments. I tried to improve the logic. I will also add the test cases for each scenario.

gatici commented on internal/juju/applications.go:
```diff
@@ -1453,44 +1453,84 @@ func (c applicationsClient) updateResources(appName string, resources map[string
+						resourceFromCharmhub.Revision = providedRev
+					}
+					pendingResourcesforAdd = append(pendingResourcesforAdd, resourceFromCharmhub)
+				} else {
+					// A new resource to be uploaded by the ResourceApi client
+					localResource := charmresources.Resource{
+						Meta:   resourceMeta,
+						Origin: charmresources.OriginUpload,
+					}
 
-	toRequest, err := resourcesAPIClient.AddPendingResources(resourcesReq)
-	if err != nil {
-		return nil, typedError(err)
+					fileSystem := osFilesystem{}
+					t, typeParseErr := charmresources.ParseType(resourceMeta.Type.String())
+					if typeParseErr != nil {
+						return nil, typedError(typeParseErr)
+					}
+					r, openResErr := resourcecmd.OpenResource(deployValue, t, fileSystem.Open)
+					if openResErr != nil {
+						return nil, typedError(openResErr)
+					}
+					toRequestUpload, err := resourcesAPIClient.UploadPendingResource(appName, localResource, deployValue, r)
+					if err != nil {
+						return nil, typedError(err)
+					}
+					// Add the resource name and the corresponding UUID to the resources map
+					toReturn[resourceMeta.Name] = toRequestUpload
+				}
+			}
+		} else {
+			// If there is no resource revisions, Charm is deployed with default resources according to channel
+			resourceFromCharmhub := charmresources.Resource{
+				Meta:     resourceMeta,
+				Origin:   charmresources.OriginStore,
+				Revision: -1,
+			}
+			pendingResourcesforAdd = append(pendingResourcesforAdd, resourceFromCharmhub)
+		}
 	}
```
thank you for refactoring the code block.  yes, it works as expected.

gatici replied:
Done

gatici commented on internal/juju/applications.go:
```diff
@@ -1453,44 +1453,84 @@ func (c applicationsClient) updateResources(appName string, resources map[string
+		// Add the resource name and the corresponding UUID to the resources map
+		for i, argsResource := range pendingResourcesforAdd {
+			toReturn[argsResource.Meta.Name] = toRequestAdd[i]
+		}
```
You are right. AddPendingResources returns toRequestAdd  which is a slice of pendingIDs and  toReturn is a  map which needs to be filled up.
Actually, this line does not override the argsResource.Meta. It fills the toReturn map using  `argsResource.Meta.Name` as a key and the pendingID comes from toRequestAdd as a value. So, nothing is overriden here.

gatici commented on internal/provider/resource_application_test.go:
```diff
@@ -344,6 +344,169 @@ func TestAcc_ResourceRevisionUpdatesMicrok8s(t *testing.T) {
+	})
+}
+
+func TestAcc_CustomResourcesRemovedFromPlanMicrok8s(t *testing.T) {
```
Done

gtato commented:
At CS we are looking forward this PR. There are some charms that upgrade their OCI images while keeping the revision unchanged, that is making pinning revisions useless. This feature would definitely help us.

gruyaume commented:
@hmlanigan could we please get a review here? We are very much looking forward to using this feature.

hmlanigan commented on internal/provider/resource_application_test.go:
```diff
@@ -669,6 +836,53 @@ resource "juju_application" "{{.AppName}}" {
+  model = juju_model.this.name
+  name = "test-app"
+  charm {
+    name     = "sdcore-ausf-k8s"
```
s/sdcore-ausf-k8s/grafana-k8s/ will fix the tests.

gatici replied:
Hello @hmlanigan, 

Integration tests and static analysis and linting issues are fixed. 
It's ready for re-review. 
Many thanks!

hmlanigan reviewed (commented):
I'm happy to land this PR once the test, static analysis and linter issues are fixed. I've added a suggestion to fix the tests.

Thank you so much for your patience and perseverance with this work!

hmlanigan reviewed (approved):
LGTM!  merging now.

hmlanigan commented:
/merge

hmlanigan commented:
The failing integration tests are timing out for taking too long. There will shortly be a PR to help with this.
